controllers/machinehealthcheck_controller.go

```diff
@@ -192,3 +265,152 @@ func (r *MachineHealthCheckReconciler) clusterToMachineHealthCheck(o handler.Map
+func (r *MachineHealthCheckReconciler) machineToMachineHealthCheck(o handler.MapObject) []reconcile.Request {
+	m, ok := o.Object.(*clusterv1.Machine)
+	if !ok {
+		r.Log.Error(errors.New("incorrect type"), "expected a Machine", "type", fmt.Sprintf("%T", o))
+		return nil
+	}
+
+	mhcList := &clusterv1.MachineHealthCheckList{}
+	if err := r.Client.List(
+		context.Background(),
+		mhcList,
+		&client.ListOptions{Namespace: m.Namespace},
+		client.MatchingFields{mhcClusterNameIndex: m.Spec.ClusterName},
+	); err != nil {
+		r.Log.Error(err, "Unable to list MachineHealthChecks", "machine", m.Name, "namespace", m.Namespace)
+		return nil
+	}
+
+	var requests []reconcile.Request
+	for k := range mhcList.Items {
+		if r.hasMatchingLabels(&mhcList.Items[k], m) {
+			key := types.NamespacedName{Namespace: mhcList.Items[k].Namespace, Name: mhcList.Items[k].Name}
+			requests = append(requests, reconcile.Request{NamespacedName: key})
+		}
+	}
+	return requests
+}
+
+func (r *MachineHealthCheckReconciler) nodeToMachineHealthCheck(o handler.MapObject) []reconcile.Request {
+	node, ok := o.Object.(*corev1.Node)
+	if !ok {
+		r.Log.Error(errors.New("incorrect type"), "expected a Node", "type", fmt.Sprintf("%T", o))
+		return nil
+	}
+
+	machine, err := r.getMachineFromNode(node.Name)
+	if machine == nil || err != nil {
+		r.Log.Error(err, "Unable to retrieve machine from node", "node", node.GetName())
+		return nil
+	}
+
+	mhcList := &clusterv1.MachineHealthCheckList{}
+	if err := r.Client.List(
+		context.Background(),
+		mhcList,
+		&client.ListOptions{Namespace: machine.Namespace},
+		client.MatchingFields{mhcClusterNameIndex: machine.Spec.ClusterName},
+	); err != nil {
+		r.Log.Error(err, "Unable to list MachineHealthChecks", "node", node.Name, "machine", machine.Name, "namespace", machine.Namespace)
+		return nil
+	}
+
+	var requests []reconcile.Request
+	for k := range mhcList.Items {
+		if r.hasMatchingLabels(&mhcList.Items[k], machine) {
+			key := types.NamespacedName{Namespace: mhcList.Items[k].Namespace, Name: mhcList.Items[k].Name}
+			requests = append(requests, reconcile.Request{NamespacedName: key})
+		}
+	}
+	return requests
+}
+
+func (r *MachineHealthCheckReconciler) getMachineFromNode(nodeName string) (*clusterv1.Machine, error) {
+	machineList := &clusterv1.MachineList{}
+	if err := r.Client.List(
+		context.TODO(),
+		machineList,
+		client.MatchingFields{machineNodeNameIndex: nodeName},
+	); err != nil {
+		return nil, errors.Wrap(err, "failed getting machine list")
+	}
+	if len(machineList.Items) != 1 {
+		return nil, errors.New(fmt.Sprintf("expecting one machine for node %v, got: %v", nodeName, machineList.Items))
+	}
+	return &machineList.Items[0], nil
+}
+
+func (r *MachineHealthCheckReconciler) watchClusterNodes(ctx context.Context, c client.Client, cluster *clusterv1.Cluster) error {
+	key := types.NamespacedName{Namespace: cluster.Name, Name: cluster.Name}
+	if _, ok := r.clusterNodeInformers[key]; ok {
```

Comment:
This probably needs locking/atomicity to support multiple workers/concurrency.

Reply:
I've added a `sync.Mutex` locking over the whole method, could possibly be better to use a `sync.RWMutex` though this might overcomplicate things maybe, WDYT?

Reply:
how about using sync.Map instead with some helper methods/funcs?

Reply:
Great idea, I've done that instead, thanks @vincepri :D